### ISSUE TITLE
wallet_transfer_to_address should accept a key as well

### ISSUE THREAD

**Issue by theoreticalbts** [BitShares/bitshares#1204]
Changing `wallet_transfer_to_address` RPC as follows:

- Change parameter type from key to string
- If string looks like a key, derive the address
- If string looks like an address, use it as-is

**Comment by theoreticalbts** [comment]
As part of the fix for this, I'm planning on refactoring `is_valid()` to return `false` when addresses are invalid instead of throwing an exception, as suggested in a TODO comment in `address.cpp`.

A breaking change of this nature in a low-level part of our API, all call sites need to be refactored to work with the new API.  To find them all, (in a branch in my own personal clone) I temporarily commented out the function's declaration and implementation, so compiler errors would allow easily finding all call sites (grepping `is_valid` got too much irrelevant stuff).

It looks like there are only two call sites.

**Comment by vikramrajkumar** [comment]
@theoreticalbts We might need to do more searching than that, I think there may be other places that don't explicitly make that call but depend on the exception throwing behavior, e.g.: https://github.com/BitShares/bitshares/blob/3cf0771afc7a70e421e91e5ee7c578e43b861213/libraries/client/wallet_api.cpp#L1033-L1054

**Comment by theoreticalbts** [comment]
TLDR, I'm not changing the contract of `address()` ctor, and most of the code that calls `is_valid()` indirectly does so through the `address()` ctor.

Longer explanation:  I'm thinking leaving `FC_ASSERT( is_valid( base58str ) )` in the constructor which would take care of that.  Anything which calls `is_valid()` indirectly must do so through code that causes `is_valid()` directly.  Of which there are two instances:

- The `address()` constructor.  In this case, the caller will still get a failed `FC_ASSERT` in every circumstance where it got an `FC_ASSERT` before the change.  The only difference is instead of asserting in `is_valid()`, `is_valid()` will return normally and the constructor will assert.

- Parameter parsing in `cli.cpp`.  In this case, I'm going to `FC_THROW` instead of FC_RETHROW`.  Again, anything that had an exception before, will still have an exception.  Just the source of exception has changed.  Since this is the CLI, it's not going to do much with the exception, it should just have a catch-all block that prints it out and returns the user to the CLI.

**Comment by vikramrajkumar** [comment]
Sounds good; forgot the ctor sort of has to throw if you try to construct a malformed addr.
